libnetdata/popen/popen.c

```diff
@@ -68,7 +68,8 @@ static inline FILE *custom_popene(const char *command, volatile pid_t *pidptr, c
     int i;
     for(i = (int) (sysconf(_SC_OPEN_MAX) - 1); i >= 0; i--)
         if(i != STDIN_FILENO && i != STDERR_FILENO)
-            fcntl(i, F_SETFD, FD_CLOEXEC);
+            if (fcntl(i, F_SETFD, FD_CLOEXEC) == -1)
+                perror("custom peopen");
```

Comment:
This was in the spirit of the [old popen](https://github.com/netdata/netdata/blob/v1.15.0/libnetdata/popen/popen.c#L73). Failing to close the files for the child process is like falling back to legacy [popen](http://man7.org/linux/man-pages/man3/popen.3.html) library call.

This is why I did not handle an error condition at all when porting the old popen to the new one. The OOM problem was because `fork()` would duplicate the address space (which could be 10s of GiB), but the per-process file descriptors being duplicated have not been observed to be a big problem yet.

I could argue that using the POSIX `popen()` semantics would be good enough and we don't even have to close any files, but there are concerns about security (leaking file descriptors to child process) and file descriptor limits.

However, the current behaviour does guarantee that a plugin will run even if it fails to close some file descriptors. So the question is, do we want to fail-stop when we fail to clean-up before cloning a process or do we tolerate file descriptor leaks so as to not disrupt netdata monitoring?

Reply:
I remember. My question was more of the rhetorical nature: is it enough to log an error? It's seldom enough or a sign of robust engineering.

I agree that it's probably not going to be a problem, but I think we should handle those things.

> So the question is, do we want to fail-stop when we fail to clean-up before cloning a process or do we tolerate file descriptor leaks so as to not disrupt netdata monitoring?

And that's something we need to discuss and agree on.

Reply:
To use the `perror` was not the best option, I agree with you, so I changed to use the `error` together with `strerror`.
Now about the original code does not do nothing when an error happens, I imagined that the original development was done based in the following sentence of the documentation  ```If the FD_CLOEXEC bit is set, the file descriptor will automatically be closed during a successful execve(2).  (If the execve(2) fails, the file descriptor is left open.) ```  , no less important , we close the PIPE few lines below this code. @mfundul can you confirm my initial thought?

Reply:
`error()` already includes `strerror` as you can see [here](https://github.com/netdata/netdata/blob/ab6f1c0b25c38e70585bfc65ec88ce5f5337f473/libnetdata/log/log.c#L758).

Moreover, this `fcntl(i, F_SETFD, FD_CLOEXEC)` is about all the file descriptors of the process, other than the pipe itself, so it's kind of the complement of the `close(pipefd[PIPE_READ]);`

Reply:
Thank you @mfundul , I fixed it.